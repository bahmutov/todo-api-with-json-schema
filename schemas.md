# Schemas


 - [postTodoRequest](#posttodorequest)
 - [postTodoResponse](#posttodoresponse)

## postTodoRequest

### postTodoRequest@1.0.0


Todo item sent by the client

name | type | required | format | description
--- | --- | --- | --- | ---
`done` | boolean | ✔ |  | Is this todo item completed?
`text` | string | ✔ |  | Todo text, like "clean room"
`uuid` | string | ✔ | [uuid](#formats) | item random GUID


Example:

```json
{
  "done": false,
  "text": "do something",
  "uuid": "20514af9-2a2a-4712-9c1e-0510c288c9ec"
}
```


[🔝](#schemas)

## postTodoResponse

### postTodoResponse@1.0.0


Todo item saved by the server and returned to the client

name | type | required | format | description
--- | --- | --- | --- | ---
`done` | boolean | ✔ |  | Is this todo item completed?
`id` | integer | ✔ |  | Item server id
`text` | string | ✔ |  | Todo text, like "clean room"
`uuid` | string | ✔ | [uuid](#formats) | item random GUID


Example:

```json
{
  "done": false,
  "id": 2,
  "text": "do something",
  "uuid": "3372137d-b582-4e32-807d-af3021112efa"
}
```


[🔝](#schemas)

## formats


Custom formats defined to better represent our data.

name | regular expression | dynamic | default
--- | --- | --- | ---
uuid | `/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/` | ✔ | `"ffffffff-ffff-ffff-ffff-ffffffffffff"`


[🔝](#schemas)

