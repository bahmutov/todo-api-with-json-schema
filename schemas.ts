import { ObjectSchema, versionSchemas } from '@cypress/schema-tools'

/**
 * Todo item sent by the client.
 */
type PostTodoRequestExample100 = {
  text: string
  done: boolean
}

const postTodoExample100: PostTodoRequestExample100 = {
  text: 'do something',
  done: false,
}

const PostTodoRequest100: ObjectSchema = {
  version: {
    major: 1,
    minor: 0,
    patch: 0,
  },
  schema: {
    title: 'PostTodoRequest',
    type: 'object',
    description: 'Todo item sent by the client',
    properties: {
      text: {
        type: 'string',
        description: 'Todo text, like "clean room"',
      },
      done: {
        type: 'boolean',
        description: 'Is this todo item completed?',
      },
    },
    additionalProperties: false,
  },
  example: postTodoExample100,
}
