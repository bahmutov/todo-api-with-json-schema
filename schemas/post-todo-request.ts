import { ObjectSchema, versionSchemas } from '@cypress/schema-tools'
import { formats } from '../formats'

type uuid = string

/**
 * Todo item sent by the client.
 */
type PostTodoRequestExample100 = {
  text: string
  done: boolean
  uuid: uuid
}

const postTodoExample100: PostTodoRequestExample100 = {
  text: 'do something',
  done: false,
  uuid: '20514af9-2a2a-4712-9c1e-0510c288c9ec',
}

const PostTodoRequest100: ObjectSchema = {
  version: {
    major: 1,
    minor: 0,
    patch: 0,
  },
  schema: {
    title: 'PostTodoRequest',
    type: 'object',
    description: 'Todo item sent by the client',
    properties: {
      text: {
        type: 'string',
        description: 'Todo text, like "clean room"',
      },
      done: {
        type: 'boolean',
        description: 'Is this todo item completed?',
      },
      uuid: {
        type: 'string',
        format: formats.uuid.name, // "uuid"
        description: 'item random GUID',
      },
    },
    // require all properties
    required: true,
    // do not allow any extra properties
    additionalProperties: false,
  },
  example: postTodoExample100,
}

export const PostTodoRequest = versionSchemas(PostTodoRequest100)
