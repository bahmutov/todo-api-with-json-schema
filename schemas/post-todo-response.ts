import { ObjectSchema, versionSchemas } from '@cypress/schema-tools'

/**
 * Todo item saved by the server and returned to the client.
 */
type PostTodoResponseExample100 = {
  text: string
  done: boolean
  id: number
}

const postTodoResponseExample100: PostTodoResponseExample100 = {
  text: 'do something',
  done: false,
  id: 2,
}

const PostTodoResponse100: ObjectSchema = {
  version: {
    major: 1,
    minor: 0,
    patch: 0,
  },
  schema: {
    title: 'PostTodoResponse',
    type: 'object',
    description: 'Todo item saved by the server and returned to the client',
    properties: {
      text: {
        type: 'string',
        description: 'Todo text, like "clean room"',
      },
      done: {
        type: 'boolean',
        description: 'Is this todo item completed?',
      },
      id: {
        type: 'integer',
        minimum: 1,
        description: 'Item server id',
      },
    },
    // require all properties
    required: true,
    // do not allow any extra properties
    additionalProperties: false,
  },
  example: postTodoResponseExample100,
}

export const PostTodoResponse = versionSchemas(PostTodoResponse100)
