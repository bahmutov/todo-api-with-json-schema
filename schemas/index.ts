import { SchemaCollection, bind, combineSchemas } from '@cypress/schema-tools'
import { formats } from '../formats'
import { PostTodoRequest } from './post-todo-request'
import { PostTodoResponse } from './post-todo-response'

export const schemas: SchemaCollection = combineSchemas(
  PostTodoRequest,
  PostTodoResponse
)

export const api = bind({ schemas, formats: formats })
/*
  api has methods to validate, sanitize, etc. objects against "schemas"
  {
    assertSchema: [Function],
    schemaNames: [ 'postTodoRequest' ],
    getExample: [Function],
    sanitize: [Function],
    validate: [Function]
  }
*/
