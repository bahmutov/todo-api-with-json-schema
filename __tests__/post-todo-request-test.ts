import { assertSchema, bind } from '@cypress/schema-tools'
import { schemas } from '../schemas'

describe('POST /todo request', () => {
  const assertTodoRequest = assertSchema(schemas)('postTodoRequest', '1.0.0')

  it('valid TODO request object', () => {
    const todo = {
      text: 'use scheams',
      done: true,
      uuid: '4899e1a9-e38f-43f9-a765-35b81a41c65d',
    }
    expect(() => {
      assertTodoRequest(todo)
    }).not.toThrow()
  })

  it('TODO request object missing text', () => {
    const todo = {
      done: true,
      uuid: '4899e1a9-e38f-43f9-a765-35b81a41c65d',
    }
    expect(() => {
      assertTodoRequest(todo)
    }).toThrowErrorMatchingSnapshot()
  })

  it('bind schemas', () => {
    const api = bind({ schemas })
    const todoRequestExample = api.getExample('postTodoRequest')('1.0.0')
    expect(todoRequestExample).toEqual({
      text: 'do something',
      done: false,
      uuid: '20514af9-2a2a-4712-9c1e-0510c288c9ec',
    })
  })

  it('bind schemas and assert an object', () => {
    const api = bind({ schemas })
    const schemaName = 'postTodoRequest'
    const schemaVersion = '1.0.0'
    const example = api.getExample(schemaName)(schemaVersion)
    const assertRequest = api.assertSchema(schemaName, schemaVersion)
    expect(() => {
      assertRequest(example)
    }).not.toThrow()
  })
})
