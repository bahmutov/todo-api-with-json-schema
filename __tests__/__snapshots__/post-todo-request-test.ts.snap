// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /todo request TODO request object missing text 1`] = `
"Schema postTodoRequest@1.0.0 violated

Errors:
data.text is required

Current object:
{
  \\"done\\": true,
  \\"uuid\\": \\"4899e1a9-e38f-43f9-a765-35b81a41c65d\\"
}

Expected object like this:
{
  \\"done\\": false,
  \\"text\\": \\"do something\\",
  \\"uuid\\": \\"20514af9-2a2a-4712-9c1e-0510c288c9ec\\"
}"
`;
